src/java/org/apache/cassandra/dht/RangeStreamer.java

```diff
@@ -373,4 +368,30 @@ public StreamResultFuture fetchAsync()
 
         return streamPlan.execute();
     }
+
+    public boolean areAllRangesPresent()
+    {
+        boolean allPresent = true;
+        for (Map.Entry<String, Map.Entry<InetAddress, Collection<Range<Token>>>> entry : toFetch.entries())
+        {
+            removeAvailableRanges(entry, false);
+            Collection<Range<Token>> ranges = entry.getValue().getValue();
+            allPresent &= ranges.isEmpty();
+        }
+        return allPresent;
+    }
+
+    @VisibleForTesting
+    void removeAvailableRanges(Map.Entry<String, Map.Entry<InetAddress, Collection<Range<Token>>>> entry, boolean log)
+    {
+        String keyspace = entry.getKey();
+        Collection<Range<Token>> ranges = entry.getValue().getValue();
```

Comment:
From reading this I'm not 100% sure if this is a reference or a copy.  I would prefer that we be more explicit and return the final set of ranges from this function.

Reply:
It's a reference since one of the unit tests would fail if just a copy. I'll refactor this to return the set!

Reply:
pushed